contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -0,0 +1,280 @@
+        TokenConfig memory config = getTokenConfigBySymbol(symbol);
+        return priceInternal(config);
+    }
+
+    function priceInternal(TokenConfig memory config) internal view returns (uint) {
+        if (config.priceSource == PriceSource.REPORTER) return prices[config.symbolHash];
+        if (config.priceSource == PriceSource.FIXED_USD) return config.fixedPrice;
+        if (config.priceSource == PriceSource.FIXED_ETH) {
+            uint usdPerEth = prices[ethHash];
+            require(usdPerEth > 0, "ETH price not set, cannot convert to dollars");
+            return mul(usdPerEth, config.fixedPrice) / config.baseUnit;
+        }
+    }
+
+    /**
+     * @notice Get the underlying price of a cToken
+     * @dev Implements the PriceOracle interface for Compound v2.
+     * @param cToken The cToken address for price retrieval
+     * @return The price for the given cToken address
+     */
+    function getUnderlyingPrice(address cToken) public view returns (uint) {
+        TokenConfig memory config = getTokenConfigByCToken(cToken);
+        return mul(1e30, priceInternal(config)) / config.baseUnit;
+    }
+
+    /**
+     * @notice Post open oracle reporter prices, and recalculate stored price by comparing to anchor
+     * @dev We let anyone pay to post anything, but only prices from configured reporter will be stored in the view.
+     * @param messages The messages to post to the oracle
+     * @param signatures The signatures for the corresponding messages
+     * @param symbols The symbols to compare to anchor for authoritative reading
+     */
+    function postPrices(bytes[] calldata messages, bytes[] calldata signatures, string[] calldata symbols) external {
+        require(messages.length == signatures.length, "messages and signatures must be 1:1");
+
+        // Save the prices
+        for (uint i = 0; i < messages.length; i++) {
+            priceData.put(messages[i], signatures[i]);
+        }
+
+        uint ethPrice = fetchEthPrice();
+
+        // Try to update the view storage
+        for (uint i = 0; i < symbols.length; i++) {
+            TokenConfig memory config = getTokenConfigBySymbol(symbols[i]);
+            string memory symbol = symbols[i];
+            bytes32 symbolHash = keccak256(abi.encodePacked(symbol));
+            if (source(messages[i], signatures[i]) != reporter) continue;
+
+            uint reporterPrice = priceData.getPrice(reporter, symbol);
+            uint anchorPrice;
+            if (symbolHash == ethHash) {
+                anchorPrice = ethPrice;
+            } else {
+                anchorPrice = fetchAnchorPrice(config, ethPrice);
+            }
+
+            uint anchorRatio = mul(anchorPrice, 100e16) / reporterPrice;
+            bool withinAnchor = anchorRatio <= upperBoundAnchorRatio && anchorRatio >= lowerBoundAnchorRatio;
+
+            if (reporterInvalidated == true) {
+                if (prices[symbolHash] == anchorPrice) {
```

Comment:
`==`?

Reply:
This isn't in the spec at all

Reply:
does solidity optimize this? eg if you are going to save the same value twice, does it still incur sstore cost?

Reply:
You get a refund with istanbul anyway right?